src/utils/signTypedData.ts

```diff
@@ -0,0 +1,30 @@
+import { _TypedDataEncoder } from '@ethersproject/hash'
+import { JsonRpcSigner } from '@ethersproject/providers'
+
+/**
+ * Overrides the _signTypedData method to add support for wallets without EIP-712 support (eg Zerion),
+ * by adding a fallback to eth_sign.
+ * @see https://github.com/ethers-io/ethers.js/blob/c80fcddf50a9023486e9f9acb1848aba4c19f7b6/packages/providers/src.ts/json-rpc-provider.ts#L334
+ */
+JsonRpcSigner.prototype._signTypedData = async function signTypedDataWithFallbacks(this, domain, types, value) {
+  // Populate any ENS names (in-place)
+  const populated = await _TypedDataEncoder.resolveNames(domain, types, value, (name: string) => {
+    return this.provider.resolveName(name) as Promise<string>
+  })
+
+  const address = await this.getAddress()
+
+  try {
+    return await this.provider.send('eth_signTypedData_v4', [
+      address.toLowerCase(),
+      JSON.stringify(_TypedDataEncoder.getPayload(populated.domain, types, populated.value)),
+    ])
+  } catch (error) {
```

Comment:
idk if really a priority, but you could technically try `eth_signTypedData` or eth_signTypedData_v3` before falling back to `eth_sign` as a general case

Reply:
we could, but in the practical case that's been tested (eg zerion) we need to resort to `eth_sign`.

Reply:
there may be other wallets that only implement the general `eth_signTypedData` case, and it's always better to use that over `eth_sign` for user safety

Reply:
Certain data may not be signable using V1/V3, but will always be signable using `eth_sign`.
Rather than adding complexity to support all versions, I think it's best for _us_ to reduce complexity/be less error prone by only supported the best case (V4) and the most general case (`eth_sign`).

Reply:
messages signed with eth_signTypedData_v3 will fail in permit2 as the contract itself verifies signatures using the v4 schema

Reply:
the docs on the differences are not great tbh lol but it seems that v3 doesn't support arrays which we need for permit2 batch permits? https://github.com/MetaMask/eth-sig-util/blob/554420a9bef69fd4307ad7a762fe81f875e57c76/src/sign-typed-data.ts#L193-L198

@snreynolds do you perchance know the diff b/t v3/v4?

Reply:
oh yeah if you need arrays then v3 won't cut it, its just that some wallets actually do v4 signing but the method name is just `eth_signTypedData` (this is an alias for general typed data signing not just v1)